File: fintech-tracker/fintechtracker-frontend/src/services/transactionService.ts
import api from "./api";

export interface Transaction {
  transactionId: number;
  userId: number;
  accountId: number;
  accountName: string;
  categoryId?: number;
  categoryName?: string;
  amount: number;
  transactionType: "income" | "expense";
  description?: string;
  transactionDate: string;
  location?: string;
  createdAt?: string;
  updatedAt?: string;
}
export interface CreateTransactionRequest {
  accountId: number;
  categoryId?: number;
  amount: number;
  transactionType: "income" | "expense";
  description?: string;
  transactionDate: string;
  location?: string;
}
export interface TransactionFilter {
  categoryId?: number;
  accountId?: number;
  fromDate?: string;
  toDate?: string;
  transactionType?: string;
  minAmount?: number;
  maxAmount?: number;
  page?: number;
  pageSize?: number;
  sortBy?: string;
  sortOrder?: string;
}
export interface TransactionSearchFilter extends TransactionFilter {
  searchTerm?: string;
}
export interface TransactionResponse {
  transactions: Transaction[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

export const transactionService = {
  async getTransactions(filter: TransactionFilter = {}) {
    try {
      // GIỮ NGUYÊN HOÀN TOÀN LOGIC XỬ LÝ PARAMS NÀY
      const params = new URLSearchParams();
      if (filter.categoryId)
        params.append("categoryId", filter.categoryId.toString());
      if (filter.accountId)
        params.append("accountId", filter.accountId.toString());
      if (filter.fromDate) params.append("fromDate", filter.fromDate);
      if (filter.toDate) params.append("toDate", filter.toDate);
      if (filter.transactionType && filter.transactionType !== "all") {
        params.append("transactionType", filter.transactionType);
      }
      if (filter.minAmount)
        params.append("minAmount", filter.minAmount.toString());
      if (filter.maxAmount)
        params.append("maxAmount", filter.maxAmount.toString());
      params.append("page", (filter.page || 1).toString());
      params.append("pageSize", (filter.pageSize || 10).toString());
      params.append("sortBy", filter.sortBy || "TransactionDate");
      params.append("sortOrder", filter.sortOrder || "desc");
      console.log("🔍 Transaction API call with params:", params.toString());

      const response = await api.get(`/transaction?${params.toString()}`);
      return response.data;
    } catch (error) {
      console.error("❌ Error fetching transactions:", error);
      throw error;
    }
  },

  async searchTransactions(searchFilter: TransactionSearchFilter = {}) {
    try {
      const { searchTerm, ...filter } = searchFilter;
      if (!searchTerm) {
        return await this.getTransactions(filter);
      }
      const params = new URLSearchParams();
      params.append("searchTerm", searchTerm);
      if (filter.categoryId)
        params.append("categoryId", filter.categoryId.toString());
      if (filter.accountId)
        params.append("accountId", filter.accountId.toString());
      if (filter.fromDate) params.append("fromDate", filter.fromDate);
      if (filter.toDate) params.append("toDate", filter.toDate);
      if (filter.transactionType && filter.transactionType !== "all") {
        params.append("transactionType", filter.transactionType);
      }
      if (filter.minAmount)
        params.append("minAmount", filter.minAmount.toString());
      if (filter.maxAmount)
        params.append("maxAmount", filter.maxAmount.toString());
      params.append("page", (filter.page || 1).toString());
      params.append("pageSize", (filter.pageSize || 10).toString());

      // SỬA ĐỔI: Thay 'apiClient' bằng 'api'
      const response = await api.get(
        `/transaction/search?${params.toString()}`
      );
      return response.data;
    } catch (error) {
      console.error("❌ Error searching transactions:", error);
      throw error;
    }
  },

  async getTransaction(id: number): Promise<Transaction> {
    const response = await api.get(`/transaction/${id}`);
    return response.data;
  },

  async createTransaction(transaction: CreateTransactionRequest) {
    const response = await api.post("/transaction", transaction);
    return response.data;
  },

  async updateTransaction(id: number, transaction: CreateTransactionRequest) {
    const response = await api.put(`/transaction/${id}`, transaction);
    return response.data;
  },

  async deleteTransaction(id: number) {
    const response = await api.delete(`/transaction/${id}`);
    return response.data;
  },
};

File: fintech-tracker/fintechtracker-backend/Services/TransactionService.cs
using fintechtracker_backend.DTOs;
using fintechtracker_backend.Repositories;

namespace fintechtracker_backend.Services
{
    public class TransactionService : ITransactionService
    {
        private readonly ITransactionRepository _transactionRepository;
        private readonly IBudgetRepository _budgetRepository;

        public TransactionService(
            ITransactionRepository transactionRepository,
            IBudgetRepository budgetRepository)
        {
            _transactionRepository = transactionRepository;
            _budgetRepository = budgetRepository;
        }

        public async Task<TransactionResponseDto> GetUserTransactionsAsync(int userId, TransactionFilterDto filter)
        {
            try
            {
                // Log incoming filter for debugging
                Console.WriteLine($"TransactionService - Received filter: CategoryId={filter.CategoryId}, AccountId={filter.AccountId}, Type={filter.TransactionType}");
                Console.WriteLine($"TransactionService - Date range: {filter.FromDate} to {filter.ToDate}");
                Console.WriteLine($"TransactionService - Amount range: {filter.MinAmount} to {filter.MaxAmount}");

                var result = await _transactionRepository.GetUserTransactionsAsync(userId, filter);

                Console.WriteLine($"TransactionService - Returned {result.Transactions.Count()} transactions");

                return result;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in TransactionService.GetUserTransactionsAsync: {ex.Message}");
                throw;
            }
        }

        public async Task<TransactionDto?> GetTransactionByIdAsync(int transactionId, int userId)
        {
            return await _transactionRepository.GetTransactionByIdAsync(transactionId, userId);
        }

        public async Task<TransactionDto> CreateTransactionAsync(int userId, CreateTransactionDto dto)
        {
            // Validate account belongs to user
            var account = await _transactionRepository.GetAccountByIdAsync(dto.AccountId, userId);
            if (account == null)
            {
                throw new UnauthorizedAccessException("Account not found or does not belong to user");
            }

            // Validate category if provided
            if (dto.CategoryId.HasValue)
            {
                var category = await _transactionRepository.GetCategoryByIdAsync(dto.CategoryId.Value);
                if (category == null)
                {
                    throw new ArgumentException("Category not found");
                }
            }

            // Create transaction
            var createdTransaction = await _transactionRepository.CreateTransactionAsync(userId, dto);

            if (createdTransaction == null)
            {
                throw new InvalidOperationException("Failed to create transaction");
            }

            return createdTransaction;
        }

        public async Task<TransactionDto?> UpdateTransactionAsync(int transactionId, int userId, UpdateTransactionDto dto)
        {
            return await _transactionRepository.UpdateTransactionAsync(transactionId, userId, dto);
        }

        public async Task<bool> DeleteTransactionAsync(int transactionId, int userId)
        {
            return await _transactionRepository.DeleteTransactionAsync(transactionId, userId);
        }

        public async Task<TransactionResponseDto> SearchTransactionsAsync(int userId, string searchTerm, TransactionFilterDto filter)
        {
            try
            {
                // Get all transactions first
                var allTransactions = await _transactionRepository.GetUserTransactionsAsync(userId, filter);

                if (string.IsNullOrEmpty(searchTerm))
                {
                    return allTransactions;
                }

                // Filter by search term (description, location, category name)
                var filteredTransactions = allTransactions.Transactions.Where(t =>
                    (!string.IsNullOrEmpty(t.Description) && t.Description.Contains(searchTerm, StringComparison.OrdinalIgnoreCase)) ||
                    (!string.IsNullOrEmpty(t.Location) && t.Location.Contains(searchTerm, StringComparison.OrdinalIgnoreCase)) ||
                    (!string.IsNullOrEmpty(t.CategoryName) && t.CategoryName.Contains(searchTerm, StringComparison.OrdinalIgnoreCase)) ||
                    (!string.IsNullOrEmpty(t.AccountName) && t.AccountName.Contains(searchTerm, StringComparison.OrdinalIgnoreCase))
                ).ToList();

                return new TransactionResponseDto
                {
                    Transactions = filteredTransactions,
                    TotalCount = filteredTransactions.Count,
                    Page = filter.Page,
                    PageSize = filter.PageSize,
                    TotalPages = (int)Math.Ceiling((double)filteredTransactions.Count / filter.PageSize)
                };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in SearchTransactionsAsync: {ex.Message}");
                throw;
            }
        }

        public async Task<List<TransactionDto>> GetTransactionsByCategoryAsync(int userId, int categoryId)
        {
            var filter = new TransactionFilterDto
            {
                CategoryId = categoryId,
                Page = 1,
                PageSize = int.MaxValue
            };

            var result = await _transactionRepository.GetUserTransactionsAsync(userId, filter);
            return result.Transactions.ToList();
        }

        public async Task<List<TransactionDto>> GetTransactionsByAccountAsync(int userId, int accountId)
        {
            var filter = new TransactionFilterDto
            {
                AccountId = accountId,
                Page = 1,
                PageSize = int.MaxValue
            };

            var result = await _transactionRepository.GetUserTransactionsAsync(userId, filter);
            return result.Transactions.ToList();
        }

        public async Task<List<TransactionDto>> GetTransactionsByDateRangeAsync(int userId, DateTime fromDate, DateTime toDate)
        {
            var filter = new TransactionFilterDto
            {
                FromDate = fromDate,
                ToDate = toDate,
                Page = 1,
                PageSize = int.MaxValue
            };

            var result = await _transactionRepository.GetUserTransactionsAsync(userId, filter);
            return result.Transactions.ToList();
        }

        public async Task<List<TransactionDto>> GetTransactionsByAmountRangeAsync(int userId, decimal minAmount, decimal maxAmount)
        {
            var filter = new TransactionFilterDto
            {
                MinAmount = minAmount,
                MaxAmount = maxAmount,
                Page = 1,
                PageSize = int.MaxValue
            };

            var result = await _transactionRepository.GetUserTransactionsAsync(userId, filter);
            return result.Transactions.ToList();
        }

        // TODO: Implement budget-related methods
        public Task<BudgetWarningDto?> CheckBudgetAfterTransactionAsync(int userId, int categoryId, decimal amount, DateTime transactionDate)
        {
            // TODO: Implement when IBudgetRepository is available
            return Task.FromResult<BudgetWarningDto?>(null);
        }

        public Task<List<BudgetAlertDto>> GetBudgetAlertsAsync(int userId)
        {
            // TODO: Implement when IBudgetRepository is available
            return Task.FromResult(new List<BudgetAlertDto>());
        }
    }
}